File: java/core/src/main/java/com/qq/tars/protocol/tars/TarsOutputStream.java
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.qq.tars.protocol.tars;

import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import com.qq.tars.common.util.HexUtil;
import com.qq.tars.protocol.tars.exc.TarsEncodeException;

public class TarsOutputStream {

    private ByteBuffer bs;

    public TarsOutputStream(ByteBuffer bs) {
        this.bs = bs;
    }

    public TarsOutputStream(int capacity) {
        bs = ByteBuffer.allocate(capacity);
    }

    public TarsOutputStream() {
        this(128);
    }

    public ByteBuffer getByteBuffer() {
        return bs;
    }

    public byte[] toByteArray() {
        byte[] newBytes = new byte[bs.position()];
        System.arraycopy(bs.array(), 0, newBytes, 0, bs.position());
        return newBytes;
    }

    public void reserve(int len) {
        if (bs.remaining() < len) {
            int n = (bs.capacity() + len) * 2;
            ByteBuffer bs2 = ByteBuffer.allocate(n);
            bs2.put(bs.array(), 0, bs.position());
            bs = bs2;
            bs2 = null;
        }
    }

    public void writeHead(byte type, int tag) {
        if (tag < 15) {
            byte b = (byte) ((tag << 4) | type);
            bs.put(b);
        } else if (tag < 256) {
            byte b = (byte) ((15 << 4) | type);
            bs.put(b);
            bs.put((byte) tag);
        } else {
            throw new TarsEncodeException("tag is too large: " + tag);
        }
    }

    public void write(boolean b, int tag) {
        byte by = (byte) (b ? 0x01 : 0);
        write(by, tag);
    }

    public void write(byte b, int tag) {
        reserve(3);
        if (b == 0) {
            writeHead(TarsStructBase.ZERO_TAG, tag);
        } else {
            writeHead(TarsStructBase.BYTE, tag);
            bs.put(b);
        }
    }

    public void write(short n, int tag) {
        reserve(4);
        if (n >= Byte.MIN_VALUE && n <= Byte.MAX_VALUE) {
            write((byte) n, tag);
        } else {
            writeHead(TarsStructBase.SHORT, tag);
            bs.putShort(n);
        }
    }

    public void write(int n, int tag) {
        reserve(6);
        if (n >= Short.MIN_VALUE && n <= Short.MAX_VALUE) {
            write((short) n, tag);
        } else {
            writeHead(TarsStructBase.INT, tag);
            bs.putInt(n);
        }
    }

    public void write(long n, int tag) {
        reserve(10);
        if (n >= Integer.MIN_VALUE && n <= Integer.MAX_VALUE) {
            write((int) n, tag);
        } else {
            writeHead(TarsStructBase.LONG, tag);
            bs.putLong(n);
        }
    }

    public void write(float n, int tag) {
        reserve(6);
        writeHead(TarsStructBase.FLOAT, tag);
        bs.putFloat(n);
    }

    public void write(double n, int tag) {
        reserve(10);
        writeHead(TarsStructBase.DOUBLE, tag);
        bs.putDouble(n);
    }

    public void writeStringByte(String s, int tag) {
        byte[] by = HexUtil.hexStr2Bytes(s);
        reserve(10 + by.length);
        if (by.length > 255) {
            writeHead(TarsStructBase.STRING4, tag);
            bs.putInt(by.length);
            bs.put(by);
        } else {
            writeHead(TarsStructBase.STRING1, tag);
            bs.put((byte) by.length);
            bs.put(by);
        }
    }

    public void writeByteString(String s, int tag) {
        reserve(10 + s.length());
        byte[] by = HexUtil.hexStr2Bytes(s);
        if (by.length > 255) {
            writeHead(TarsStructBase.STRING4, tag);
            bs.putInt(by.length);
            bs.put(by);
        } else {
            writeHead(TarsStructBase.STRING1, tag);
            bs.put((byte) by.length);
            bs.put(by);
        }
    }

    public void write(String s, int tag) {
        byte[] by;
        try {
            by = s.getBytes(sServerEncoding);
        } catch (UnsupportedEncodingException e) {
            by = s.getBytes();
        }
        reserve(10 + by.length);
        if (by.length > 255) {
            writeHead(TarsStructBase.STRING4, tag);
            bs.putInt(by.length);
            bs.put(by);
        } else {
            writeHead(TarsStructBase.STRING1, tag);
            bs.put((byte) by.length);
            bs.put(by);
        }
    }

    public <K, V> void write(Map<K, V> m, int tag) {
        reserve(8);
        writeHead(TarsStructBase.MAP, tag);
        write(m == null ? 0 : m.size(), 0);
        if (m != null) {
            for (Map.Entry<K, V> en : m.entrySet()) {
                write(en.getKey(), 0);
                write(en.getValue(), 1);
            }
        }
    }

    public void write(boolean[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (boolean e : l)
            write(e, 0);
    }

    public void write(byte[] l, int tag) {
        reserve(8 + l.length);
        writeHead(TarsStructBase.SIMPLE_LIST, tag);
        writeHead(TarsStructBase.BYTE, 0);
        write(l.length, 0);
        bs.put(l);
    }

    public void write(short[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (short e : l)
            write(e, 0);
    }

    public void write(int[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (int e : l)
            write(e, 0);
    }

    public void write(long[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (long e : l)
            write(e, 0);
    }

    public void write(float[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (float e : l)
            write(e, 0);
    }

    public void write(double[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (double e : l)
            write(e, 0);
    }

    public <T> void write(T[] l, int tag) {
        writeArray(l, tag);
    }

    private void writeArray(Object[] l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l.length, 0);
        for (Object e : l)
            write(e, 0);
    }

    public <T> void write(Collection<T> l, int tag) {
        reserve(8);
        writeHead(TarsStructBase.LIST, tag);
        write(l == null ? 0 : l.size(), 0);
        if (l != null) {
            for (T e : l)
                write(e, 0);
        }
    }

    public void write(TarsStructBase o, int tag) {
        reserve(2);
        writeHead(TarsStructBase.STRUCT_BEGIN, tag);
        o.writeTo(this);
        reserve(2);
        writeHead(TarsStructBase.STRUCT_END, 0);
    }

    public void write(Byte o, int tag) {
        write(o.byteValue(), tag);
    }

    public void write(Boolean o, int tag) {
        write(o.booleanValue(), tag);
    }

    public void write(Short o, int tag) {
        write(o.shortValue(), tag);
    }

    public void write(Integer o, int tag) {
        write(o.intValue(), tag);
    }

    public void write(Long o, int tag) {
        write(o.longValue(), tag);
    }

    public void write(Float o, int tag) {
        write(o.floatValue(), tag);
    }

    public void write(Double o, int tag) {
        write(o.doubleValue(), tag);
    }

    public void write(Object o, int tag) {
        if (o instanceof Byte) {
            write(((Byte) o).byteValue(), tag);
        } else if (o instanceof Boolean) {
            write(((Boolean) o).booleanValue(), tag);
        } else if (o instanceof Short) {
            write(((Short) o).shortValue(), tag);
        } else if (o instanceof Integer) {
            write(((Integer) o).intValue(), tag);
        } else if (o instanceof Long) {
            write(((Long) o).longValue(), tag);
        } else if (o instanceof Float) {
            write(((Float) o).floatValue(), tag);
        } else if (o instanceof Double) {
            write(((Double) o).doubleValue(), tag);
        } else if (o instanceof String) {
            write((String) o, tag);
        } else if (o instanceof Map) {
            write((Map<?, ?>) o, tag);
        } else if (o instanceof List) {
            write((List<?>) o, tag);
        } else if (o instanceof TarsStructBase) {
            write((TarsStructBase) o, tag);
        } else if (o instanceof byte[]) {
            write((byte[]) o, tag);
        } else if (o instanceof boolean[]) {
            write((boolean[]) o, tag);
        } else if (o instanceof short[]) {
            write((short[]) o, tag);
        } else if (o instanceof int[]) {
            write((int[]) o, tag);
        } else if (o instanceof long[]) {
            write((long[]) o, tag);
        } else if (o instanceof float[]) {
            write((float[]) o, tag);
        } else if (o instanceof double[]) {
            write((double[]) o, tag);
        } else if (o.getClass().isArray()) {
            writeArray((Object[]) o, tag);
        } else if (o instanceof Collection) {
            write((Collection<?>) o, tag);
        } else {
            TarsOutputStreamExt.write(o, tag, this);
        }
    }

    protected String sServerEncoding = "UTF-8";

    public int setServerEncoding(String se) {
        sServerEncoding = se;
        return 0;
    }

    public static void main(String[] args) {
        TarsOutputStream os = new TarsOutputStream();
        long n = 0x1234567890012345L;
        os.write(n, 0);
        ByteBuffer bs = os.getByteBuffer();
        System.out.println(HexUtil.bytes2HexStr(bs.array()));
        System.out.println(Arrays.toString(os.toByteArray()));
    }
}

File: tup/tup-csharp/tars/TarsOutputStream.cs
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except 
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed 
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR 
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the 
 * specific language governing permissions and limitations under the License.
 */


using Tup.Tars;
using Tup;

using System;
using System.Text;
using System.IO;
using System.Collections.Generic;
using System.Collections;

namespace Tup.Tars
{
    public class TarsOutputStream
    {
        private MemoryStream ms;
        private BinaryWriter bw;

        public TarsOutputStream(MemoryStream ms)
        {
            this.ms = ms;
            bw = new BinaryWriter(ms, Encoding.BigEndianUnicode);
        }

        public TarsOutputStream(int capacity)
        {
            ms = new MemoryStream(capacity);
            bw = new BinaryWriter(ms, Encoding.BigEndianUnicode);
        }

        public TarsOutputStream()
        {
            ms = new MemoryStream(128);
            bw = new BinaryWriter(ms, Encoding.BigEndianUnicode);
        }

        public MemoryStream getMemoryStream()
        {
            return ms;
        }

        public byte[] toByteArray()
        {
            byte[] newBytes = new byte[ms.Position];
            System.Array.Copy(ms.GetBuffer(), 0, newBytes, 0, (int)ms.Position);
            return newBytes;
        }

        public void reserve(int len)
        {
            int nRemain = ms.Capacity - (int)ms.Length;
            if (nRemain < len)
            {
                ms.Capacity = (ms.Capacity + len) << 1;// -nRemain;
            }
        }

        // 写入头信息
        public void writeHead(byte type, int tag)
        {
            if (tag < 15)
            {
                byte b = (byte)((tag << 4) | type);

                try
                {
                    {
                        bw.Write(b);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }

            }
            else if (tag < 256)
            {
                try
                {
                    byte b = (byte)((15 << 4) | type);
                    {
                        bw.Write(b);
                        bw.Write((byte)tag);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(this + " writeHead: " + e.Message);
                }
            }
            else
            {
                throw new TarsEncodeException("tag is too large: " + tag);
            }
        }

        public void Write(bool b, int tag)
        {
            byte by = (byte)(b ? 0x01 : 0);
            Write(by, tag);
        }

        public void Write(byte b, int tag)
        {
            reserve(3);
            if (b == 0)
            {
                writeHead((byte)TarsStructType.ZERO_TAG, tag);
            }
            else
            {
                writeHead((byte)TarsStructType.BYTE, tag);
                try
                {
                    {
                        bw.Write(b);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
        }

        public void Write(short n, int tag)
        {
            reserve(4);
            if(n>=-128 && n<=127)
            {
                Write((byte)n, tag);
            }
            else
            {
                writeHead((byte)TarsStructType.SHORT, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(n));
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(this + " Write: " + e.Message);
                }
            }
        }
        public void Write(ushort n, int tag)
        {
            Write((short)n, tag);
        }

        public void Write(int n, int tag)
        {
            reserve(6);

            if (n >= short.MinValue && n <= short.MaxValue)
            {
                Write((short)n, tag);
            }
            else
            {
                writeHead((byte)TarsStructType.INT, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(n));
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
        }

        public void Write(uint n, int tag)
        {
            Write((long)n, tag);
        }

        public void Write(ulong n, int tag)
        {
            Write((long)n, tag);
        }

        public void Write(long n, int tag)
        {
            reserve(10);
            if (n >= int.MinValue && n <= int.MaxValue)
            {
                Write((int)n, tag);
            }
            else
            {
                writeHead((byte)TarsStructType.LONG, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(n));
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
        }

        public void Write(float n, int tag)
        {
            reserve(6);
            writeHead((byte)TarsStructType.FLOAT, tag);
            try
            {
                {
                    bw.Write(ByteConverter.ReverseEndian(n));
                }
            }
            catch (Exception e)
            {
                QTrace.Trace(e.Message);
            }
        }

        public void Write(double n, int tag)
        {
            reserve(10);
            writeHead((byte)TarsStructType.DOUBLE, tag);
            try
            {
                {
                    bw.Write(ByteConverter.ReverseEndian(n));
                }
            }
            catch (Exception e)
            {
                QTrace.Trace(e.Message);
            }
        }

        public void writeStringByte(string s, int tag)
        {
            byte[] by = HexUtil.hexStr2Bytes(s);
            reserve(10 + by.Length);
            if (by.Length > 255)
            {
                // 长度大于255，为String4类型
                writeHead((byte)TarsStructType.STRING4, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(by.Length));
                        bw.Write(by);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
            else
            {
                // 长度小于255，位String1类型
                writeHead((byte)TarsStructType.STRING1, tag);
                try
                {
                    {
                        bw.Write((byte)by.Length);
                        bw.Write(by);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
        }

        public void writeByteString(string s, int tag)
        {
            reserve(10 + s.Length);
            byte[] by = HexUtil.hexStr2Bytes(s);
            if (by.Length > 255)
            {
                writeHead((byte)TarsStructType.STRING4, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(by.Length));
                        bw.Write(by);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
            else
            {
                writeHead((byte)TarsStructType.STRING1, tag);
                try
                {
                    {
                        bw.Write((byte)by.Length);
                        bw.Write(by);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
        }

        public void Write(string s, int tag,bool IsLocalString = false)
        {
            byte[] by;
            try
            {
                by = ByteConverter.String2Bytes(s, IsLocalString);
                if (by == null)
                {
                    by = new byte[0];
                }
            }
            catch (Exception e)
            {
                Tup.QTrace.Trace(this + " write s Exception" + e.Message);
                return;
            }
            if (by != null)
            {
                reserve(10 + by.Length);
            }
            if (by != null && by.Length > 255)
            {
                writeHead((byte)TarsStructType.STRING4, tag);
                try
                {
                    {
                        bw.Write(ByteConverter.ReverseEndian(by.Length));
                        bw.Write(by);
                    }
                }
                catch (Exception e)
                {
                    QTrace.Trace(e.Message);
                }
            }
            else
            {
                writeHead((byte)TarsStructType.STRING1, tag);
                try
                {
                    {
                        if (by != null)
                        {
                            bw.Write((byte)by.Length);
                            bw.Write(by);
                        }
                        else
                        {
                            bw.Write((byte)0);
                        }
                    }
                }
                catch (Exception e)
                {
                    Tup.QTrace.Trace(this + " write s(2) Exception" + e.Message);
                }
            }
        }

        public void write<K, V>(Dictionary<K, V> m, int tag)
        {
            reserve(8);
            writeHead((byte)TarsStructType.MAP, tag);
            
            Write(m == null ? 0 : m.Count, 0);
            if (m != null)
            {
                foreach (KeyValuePair<K, V> en in m)
                {
                    Write(en.Key, 0);
                    Write(en.Value, 1);
                }
            }
        }

        public void Write(IDictionary m, int tag)
        {
            reserve(8);
            writeHead((byte)TarsStructType.MAP, tag);
            Write(m == null ? 0 : m.Count, 0);
            if (m != null)
            {
                ICollection keys = m.Keys;
                foreach (object objKey in keys)
                {
                    Write(objKey, 0);
                    Write(m[objKey], 1);
                }
            }
        }

        public void Write(bool[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);

            if (l != null)
            {
                foreach (bool e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void Write(byte[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8 + nLen);
            writeHead((byte)TarsStructType.SIMPLE_LIST, tag);
            writeHead((byte)TarsStructType.BYTE, 0);
            Write(nLen, 0);

            try
            {
                if(l != null)
                {
                    bw.Write(l);
                }
            }
            catch (Exception e)
            {
                QTrace.Trace(e.Message);
            }
        }

        public void Write(short[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);
            if (l != null)
            {
                foreach (short e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void Write(int[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);
            if (l != null)
            {
                foreach (int e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void Write(long[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);

            if (l != null)
            {
                foreach (long e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void Write(float[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);
            if (l != null)
            {
                foreach (float e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void Write(double[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);

            if (l != null)
            {
                foreach (double e in l)
                {
                    Write(e, 0);
                }
            }
        }

        public void write<T>(T[] l, int tag)
        {
            object o = l;
            writeArray((object[])o, tag);
        }

        private void writeArray(object[] l, int tag)
        {
            int nLen = 0;
            if (l != null)
            {
                nLen = l.Length;
            }
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(nLen, 0);

            if (l != null)
            {
                foreach (Object e in l)
                {
                    Write(e, 0);
                }
            }
        }

        // 由于list，应该在第一个位置[0]预置一个元素（可以为空），以方便判断元素类型
        public void writeList(IList l, int tag)
        {
            reserve(8);
            writeHead((byte)TarsStructType.LIST, tag);
            Write(l == null ? 0 : (l.Count>0?l.Count:0), 0);
            if (l != null)
            {
                for(int i = 0; i<l.Count; i++)
                {
                    Write(l[i], 0);
                }
            }
        }

        public void Write(TarsStruct o, int tag)
        {
            if (o == null)
            {
                return;
            }
            reserve(2);
            writeHead((byte)TarsStructType.STRUCT_BEGIN, tag);
            o.WriteTo(this);
            reserve(2);
            writeHead((byte)TarsStructType.STRUCT_END, 0);
        }

        public void Write(object o, int tag)
        {
            if (o == null)
            {
                return;
            }
            if (o is byte)
            {
                Write(((byte)o), tag);
            }
            else if (o is Boolean)
            {
                Write((bool)o, tag);
            }
            else if (o is short)
            {
                Write(((short)o), tag);
            }
            else if (o is ushort)
            {
                Write(((int)(ushort)o), tag);
            }
            else if (o is int)
            {
                Write(((int)o), tag);
            }
            else if (o is uint)
            {
                Write((long)(uint)o, tag);
            }
            else if (o is long)
            {
                Write(((long)o), tag);
            }
            else if (o is ulong)
            {
                Write(((long)(ulong)o), tag);
            }
            else if (o is float)
            {
                Write(((float)o), tag);
            }
            else if (o is double)
            {
                Write(((double)o), tag);
            }
            else if (o is string)
            {
                string strObj = o as string;
                Write(strObj, tag);
            }
            else if (o is TarsStruct)
            {
                Write((TarsStruct)o, tag);
            }
            else if (o is byte[])
            {
                Write((byte[])o, tag);
            }
            else if (o is bool[])
            {
                Write((bool[])o, tag);
            }
            else if (o is short[])
            {
                Write((short[])o, tag);
            }
            else if (o is int[])
            {
                Write((int[])o, tag);
            }
            else if (o is long[])
            {
                Write((long[])o, tag);
            }
            else if (o is float[])
            {
                Write((float[])o, tag);
            }
            else if (o is double[])
            {
                Write((double[])o, tag);
            }
            else if (o.GetType().IsArray)
            {
                Write((object[])o, tag);
            }
            else if (o is IList)
            {
                writeList((IList)o, tag);
            }
            else if (o is IDictionary)
            {
                Write((IDictionary)o, tag);
            }
            else
            {
                throw new TarsEncodeException(
                        "write object error: unsupport type. " + o.ToString() + "\n");
            }
        }

        protected string sServerEncoding = "UTF-8";
        public int setServerEncoding(string se)
        {
            sServerEncoding = se;
            return 0;
        }
    }
}
